builtin/providers/cloudstack/resource_cloudstack_affinity_group.go

```diff
@@ -0,0 +1,136 @@
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceCloudStackAffinityGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	name := d.Get("name").(string)
+	affinityGroupType := d.Get("type").(string)
+
+	log.Printf("[DEBUG] creating affinity group with name %s of type %s", name, affinityGroupType)
+
+	p := cs.AffinityGroup.NewCreateAffinityGroupParams(name, affinityGroupType)
+
+	// Set the description
+	description, hasDescription := d.GetOk("description")
+	if hasDescription {
+		p.SetDescription(description.(string))
+	}
+
+	// If there is a project supplied, we retrieve and set the project id
+	if err := setProjectid(p, cs, d); err != nil {
+		return err
+	}
+
+	r, err := cs.AffinityGroup.CreateAffinityGroup(p)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] New affinity group successfully generated at Cloudstack")
+	d.SetId(r.Id)
+
+	return resourceCloudStackAffinityGroupRead(d, meta)
+}
+
+func resourceCloudStackAffinityGroupRead(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	log.Printf("[DEBUG] looking for affinity group with name %s", d.Id())
+
+	p := cs.AffinityGroup.NewListAffinityGroupsParams()
+	p.SetId(d.Id())
+
+	// If there is a project supplied, we retrieve and set the project id
+	if err := setProjectid(p, cs, d); err != nil {
+		return err
+	}
+
+	r, err := cs.AffinityGroup.ListAffinityGroups(p)
+	if err != nil {
+		return err
+	}
+	if r.Count == 0 {
+		log.Printf("[DEBUG] Affinity group %s does not exist", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	//Affinity group name is unique in a cloudstack account so dont need to check for multiple
+	d.Set("name", r.AffinityGroups[0].Name)
+	d.Set("description", r.AffinityGroups[0].Description)
+	d.Set("type", r.AffinityGroups[0].Type)
+
+	return nil
+}
+
+func resourceCloudStackAffinityGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	// Create a new parameter struct
+	p := cs.AffinityGroup.NewDeleteAffinityGroupParams()
+
+	// Set id
+	p.SetId(d.Id())
+
+	// If there is a project supplied, we retrieve and set the project id
+	if err := setProjectid(p, cs, d); err != nil {
+		return err
+	}
+
+	// Remove the affinity group
+	_, err := cs.AffinityGroup.DeleteAffinityGroup(p)
+	if err != nil {
+		// This is a very poor way to be told the ID does no longer exist :(
+		if strings.Contains(err.Error(), fmt.Sprintf(
+			"Unable to find affinity group %s", d.Id())) {
+			return nil
```

Comment:
Does the API for this call differ from the other API responses? As the more generic way to check for this (which is used in all other places) is:

```go
if strings.Contains(err.Error(), fmt.Sprintf(
  "Invalid parameter id value=%s due to incorrect long value format, "+
    "or entity does not exist", d.Id())) {
  return nil
}
```